File: utils/tmux.go
package utils

import (
	"fmt"
	"os"
	"os/exec"
	"strings"
	"syscall"
)

func InTmux() bool {
	_, isSet := os.LookupEnv("TMUX")
	return isSet
}

func TmuxActive() (bool, error) {

	if InTmux() {
		return true, nil
	}

	cmd := exec.Command("tmux", "info", "-t", "/dev/pts/1")
	res, _ := cmd.CombinedOutput()

	if strings.Contains(string(res), "no server") {
		return false, nil
	} else {
		return true, nil
	}

	return false, nil
}

func TmuxSessionExists(sessionName string) (bool, error) {

	// If Tmux is Not Active, Then the Session Can't Exist
	active, err := TmuxActive()
	if err != nil {
		return false, err
	}

	if active {

		cmd := exec.Command("tmux", "has-session", "-t", sessionName)
		res, err := cmd.CombinedOutput()

		if err != nil {
			return false, nil
		}

		if strings.Contains(string(res), "can't find session") {
			return false, nil
		} else {
			return true, nil
		}

	}

	return false, nil
}

func TmuxWindowExists(sessionName string, windowName string) (bool, error) {

	// In Order for The Window to Exist, the Session Must Aswell
	sessionExists, err := TmuxSessionExists(sessionName)
	if err != nil {
		return false, nil
	}

	if sessionExists {

		cmd := exec.Command("tmux", "has-session", "-t", sessionName+":"+windowName)
		res, _ := cmd.CombinedOutput()
		if err != nil {
			return false, nil
		}

		if strings.Contains(string(res), "can't find window") {
			return false, nil
		} else {
			return true, nil
		}

	}
	return false, nil
}

func TmuxCreateSession(sessionName string) (bool, error) {

	// Do not recreate the session if one exists
	sessionExists, err := TmuxSessionExists(sessionName)
	if err != nil {

		return false, err
	}

	if sessionExists {
		return true, nil

	} else {
		// if session does not exist, create Session
		cmd := exec.Command("tmux", "new-session", "-d", "-s", sessionName)
		_, err := cmd.CombinedOutput()

		if err != nil {
			return false, err
		}
	}
	return true, nil
}

func TmuxCreateWindow(sessionName string, windowName string) (bool, error) {

	// If Window Exists, do not create window
	windowExists, err := TmuxWindowExists(sessionName, windowName)
	if err != nil {
		return false, err
	}

	// If Session Exists, do not recreate Session
	sessionExists, err := TmuxSessionExists(sessionName)
	if err != nil {
		return false, err
	}

	if windowExists {
		return true, nil
	} else if sessionExists {
		cmd := exec.Command("tmux", "new-window", "-t", sessionName, "-n", windowName)
		_, err := cmd.CombinedOutput()
		if err != nil {
			return false, err
		}

	} else {
		cmd := exec.Command("tmux", "new-session", "-d", "-s", sessionName, "-n", windowName)
		_, err := cmd.CombinedOutput()

		if err != nil {
			return false, err
		}

		return true, nil
	}

	return false, nil

}

func TmuxAttachOrSelectWindow(sessionName string, windowName string) (bool, error) {

	windowExists, err := TmuxCreateWindow(sessionName, windowName)

	if err != nil {
		return false, err
	}

	if windowExists && InTmux() {
		tmux, _ := exec.LookPath("tmux")
		args := []string{"tmux", "switch", "-t", sessionName, ":", windowName}
		err := syscall.Exec(tmux, args, os.Environ())
		if err != nil {
			return false, err
		}

		return true, nil
	} else if windowExists && !InTmux() {
		tmux, _ := exec.LookPath("tmux")
		args := []string{"tmux", "attach", "-t", sessionName}
		err := syscall.Exec(tmux, args, os.Environ())
		if err != nil {
			return false, err
		}
		return true, nil
	}

	return false, nil
}

func TmuxSendKeys(sessionName, windowName string, sendKeys string) (bool, error) {

	cmd := exec.Command("tmux", "send-keys", "-t", sessionName+":"+windowName, sendKeys, "C-m")
	_, err := cmd.CombinedOutput()

	if err != nil {
		return false, err
	} else {
		return true, nil
	}

	return false, nil

}

func TmuxListWindows() (bool, error) {

	cmd := exec.Command("tmux", "list-windows")
	output, err := cmd.CombinedOutput()

	fmt.Println(string(output))

	windows := string.split(output, "\n")

  for window := range(windows) {

  }

	if err != nil {
		return false, err
	} else {
		return true, nil
	}

	return false, nil

}

func TmuxSplitWindow(sessionName string, windowName string) (bool, error) {

	cmd := exec.Command("tmux", "split-window")
	_, err := cmd.CombinedOutput()

	if err != nil {
		return false, err
	} else {
		return true, nil
	}

	return false, nil

}

File: src/tmux.rs
use std::{env, process, str};
pub struct TMUX {}

impl TMUX {
    pub fn in_tmux() -> bool {
        return env::var("TMUX").is_ok();
    }

    pub fn run_tmux_command(args: Vec<&str>) -> bool {
        let status = process::Command::new("tmux")
            .args(args)
            .stdout(process::Stdio::null())
            .stderr(process::Stdio::null())
            .status()
            .unwrap();
        return status.success();
    }

    pub fn active() -> bool {
        if Self::in_tmux() {
            return true;
        }

        let output: process::Output = process::Command::new("tmux")
            .args(vec!["-t", "/dev/pts/1"])
            .output()
            .unwrap()
            .into();

        if str::from_utf8(&output.stdout)
            .unwrap()
            .contains("no server")
        {
            return false;
        } else {
            return true;
        }
    }

    pub fn session_exists(session_name: &str) -> bool {
        // If tmux is not active, session_name cant exist
        if !Self::active() {
            return false;
        }

        return Self::run_tmux_command(vec!["has-session", "-t", session_name]);
    }

    pub fn window_exists(session_name: &str, window_name: &str) -> bool {
        // In order for window to exist, session must exist
        if !Self::session_exists(session_name) {
            return false;
        }

        return Self::run_tmux_command(vec![
            "has-session",
            "-t",
            &format!("{session_name}:{window_name}"),
        ]);
    }

    pub fn create_window(session_name: &str, window_name: &str) -> bool {
        // Do not recreate window if one exists
        if Self::window_exists(session_name, window_name) {
            return true;
        }

        // If session exists, do not recreate Session
        if Self::session_exists(session_name) {
            return Self::run_tmux_command(vec![
                "new-window",
                "-t",
                session_name,
                "-n",
                window_name,
            ]);
        }

        // Create Session and Window in Detached State
        return Self::run_tmux_command(vec![
            "new-session",
            "-d",
            "-s",
            session_name,
            "-n",
            window_name,
        ]);
    }

    pub fn attach_or_select_window(session_name: &str, window_name: &str) -> bool {
        if !Self::window_exists(session_name, window_name) {
            panic!("Window Doesnt Exist!, Please create window first!");
        }

        if Self::in_tmux() {
            let _o = process::Command::new("tmux")
                .args(vec![
                    "switch",
                    "-t",
                    &format!("{session_name}:{window_name}"),
                ])
                .spawn()
                .expect("FAILED TO ATTACH")
                .wait();
            return true;
        } else {
            let _o = process::Command::new("tmux")
                .args(vec![
                    "attach",
                    "-t",
                    &format!("{session_name}:{window_name}"),
                ])
                .spawn()
                .expect("FAILED TO ATTACH")
                .wait();
            return true;
        }
    }

    pub fn send_keys(session_name: &str, window_name: &str, keys: &str) -> bool {
        return Self::run_tmux_command(vec![
            "send-keys",
            "-t",
            &format!("{session_name}:{window_name}"),
            keys,
            "C-m",
        ]);
    }

    pub fn watch_command(
        session_name: &str,
        window_name: &str,
        hidden: bool,
        horizontal: bool,
        command: &str,
    ) -> bool {
        let dir_arg: &str;
        if horizontal {
            dir_arg = "-h";
        } else {
            dir_arg = "-v";
        }

        if hidden {
            return Self::run_tmux_command(vec![
                "split-window",
                "-d",
                &dir_arg,
                "-t",
                &format!("{session_name}:{window_name}"),
                "-Z",
                "watch",
                &command,
            ]);
        } else {
            return Self::run_tmux_command(vec![
                "split-window",
                "-d",
                &dir_arg,
                "-t",
                &format!("{session_name}:{window_name}"),
                "watch",
                &command,
            ]);
        }

        return false;
    }
}
